robottime: Honor the SOURCE_DATE_EPOCH environment variable.

Apteryks opened the pull request:
* src/robot/utils/robottime.py:
(TimestampCache._get_epoch): Retrieve date from SOURCE_DATE_EPOCH if
it is defined, otherwise from time.time().
* utest/output/test_logger.py (TestLogger.test_write_to_one_logger):
Verify the timestamp via a regular expression.
(TestLogger.test_write_to_logger_with_source_date_epoch_set): New test.
* doc/userguide/src/Appendices/ReproducibleBuilds.rst: New file.
* doc/userguide/src/RobotFrameworkUserGuide.rst: Register it.

pekkaklarck commented on src/robot/utils/robottime.py:
```diff
@@ -420,6 +421,8 @@ def get_timestamp(self, daysep='', daytimesep=' ', timesep=':', millissep='.'):
     def _get_epoch(self):
+        if os.getenv('SOURCE_DATE_EPOCH'):
+            return float(os.getenv('SOURCE_DATE_EPOCH'))
         return time.time()
```
This is pretty performance critical code, timestamps are generated for each keyword, loop, etc. both when they start and end, and thus always using `os.getenv` isn't great. It would be better to have the `if` outside the `_get_epoch` like this:

```python
if 'SOURCE_DATE_EPOCH' in os.environ:
    def _get_epoch(self):
        return float(os.getenv('SOURCE_DATE_EPOCH'))
else:
    def _get_epoch(self):
        return time.time()

Apteryks replied:
> Could you @Apteryks still clarify the use case for this change a bit?

My specific use case was just fixing the timestamps embedded in the generated documentation so that RobotFramework (and its libraries) could be built reproducibly in GNU Guix. Another, more contrived example, could be the desire that two RobotFramework test runs produce exactly the same output.

Apteryks replied:
I've made the suggested change, and it expectedly broke the new `test_write_to_logger_with_source_date_epoch_set`; which I tried fixing by mocking the `TIMESTAMP_CACHE` global object:
```diff
modified   utest/output/test_logger.py
@@ -4,6 +4,7 @@ import time
 import unittest
 import unittest.mock
 
+import robot.utils.robottime
 from robot.utils.asserts import assert_equal, assert_true, assert_false
 from robot.utils.robottime import format_time
 
@@ -59,6 +60,8 @@ class TestLogger(unittest.TestCase):
                              logger.msg.timestamp))
 
     @unittest.mock.patch.dict(os.environ, {'SOURCE_DATE_EPOCH': "0"})
+    @unittest.mock.patch('robot.utils.robottime.TIMESTAMP_CACHE',
+                         robot.utils.robottime.TimestampCache())
     def test_write_to_logger_with_source_date_epoch_set(self):
         logger = LoggerMock(('Hello, SOURCE_DATE_EPOCH!', 'INFO'))
         self.logger.register_logger(logger)
```
But it didn't work.  Any suggestion?

pekkaklarck replied:
Yeah, `if 'SOURCE_DATE_EPOCH' in os.environ:` in my proposal is executed when the `robottime` module is initially imported and thus patching it isn't easy. The code is so trivial that perhaps we wouldn't need a test at all. There could be just a comment explaining why `SOURCE_DATE_EPOCH` is used with an URL pointing to the original issue.

pekkaklarck commented on doc/userguide/src/Appendices/ReproducibleBuilds.rst:
```diff
@@ -0,0 +1,9 @@
+variable so that it can be built in a reproducible fashion.  When the
+`SOURCE_DATE_EPOCH` environment variable is set, timestamps embedded
+in generated documentation and log messages are set to its fixed
+value.
```
Would BUILD.rst be better place for this information than the User Guide?

Apteryks replied:
BUILD.rst seems to be specific to creating Robot Framework releases.  The `SOURCE_DATE_EPOCH` environment variable is most useful for the (automated) build tooling of 3rd parties packaging RobotFramework such as GNU Guix, NixOS or Debian that strive for reproducible builds.  So it doesn't really need to be user-facing, but it's also not at home in BUILD.rst.  I'm not sure of the value to document it, but if we do, an appendix seems the best place to me (it's there if someone looks for it, but it's in the background as most users probably do not care).

pekkaklarck replied:
Isn't `SOURCE_DATE_EPOCH` exactly about creating releases? I don't see problems enhancing BUILD.rst to cover also third party packaging.

pekkaklarck commented:
Could you @Apteryks still clarify the use case for this change a bit? I understand the general idea behind the `SOURCE_DATE_EPOCH` environment variable, but I don't fully understand how it would be used with Robot. All timestamps reading this environment value instead of using `time.time()` means that all tests, keywords, etc. in log have that time both as their start and end time. That doesn't sound too useful, and I was wondering should the environment variable be checked in some other place that where this PR now checks it.

I try to get RF 5.0 rc 1 finally out on Thursday so there's a bit hurry in getting this in. Feel free to ping me on Slack, preferably on the #devel channel so that others can also participate the discussion, to have more interactive discussion.

pekkaklarck commented:
I'm starting to believe `robottime` isn't the right place to look at `SOURCE_DATE_EPOCH` at all. See my recent comments to issue #4262 for details.

Apteryks commented:
This was closed by mistake, due to recreating the branch (sorry!).  I believe I've addressed the comments thus far in the newly opened https://github.com/robotframework/robotframework/pull/4286.
